File: 934. Shortest Bridge
/Solution.cpp
class Solution {
public:
    int r, c;
    queue<vector<int>> q;
    vector<vector<int>> g;
    vector<vector<int>> moves = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    vector<vector<bool>> visited;
    int shortestBridge(vector<vector<int>>& grid) {
        r = grid.size();
        c = grid[0].size();
        g = grid;
        for (int i = 0; i < r; i++) {
            vector<bool> s(c, false);
            visited.push_back(s);
        }
        int count = 0;
        for (int i = 0; i < r && q.empty(); i++) {
            for (int j = 0; j < c && q.empty(); j++) {
                if (grid[i][j] == 0) continue;
                find(i, j);
            }
        }
        while (!q.empty()) {
            int n = q.size();
            for (int i = 0; i < n; i++) {
                vector<int> curr = q.front();
                q.pop();
                for (vector<int> m : moves) {
                    int nx = curr[0] + m[0], ny = curr[1] + m[1];
                    if (nx < 0 || ny < 0 || nx >= r || ny >= c) continue;
                    if (visited[nx][ny]) continue;
                    if (grid[nx][ny] == 1) return count;
                    visited[nx][ny] = true;
                    vector<int> foll = { nx, ny };
                    q.push(foll);
                }
            }
            count++;
        }
        return -1;
    }

    void find(int x, int y) {
        if (x < 0 || y < 0 || x >= r || y >= c) return;
        if (visited[x][y] || g[x][y] == 0) return;
        visited[x][y] = true;
        vector<int> foll = { x, y };
        q.push(foll);
        for (vector<int> m : moves) {
            find(x + m[0], y + m[1]);
        }
    }
};

File: 934. Shortest Bridge
/Solution.py
class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        self.r, self.c = len(grid), len(grid[0])
        self.q, self.grid, self.visited = [], grid, [[False for i in range(self.c)] for j in range(self.r)]
        self.moves = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        count = 0
        for i in range(self.r):
            for j in range(self.c):
                if grid[i][j] == 0: continue
                self.find(i, j)
                break
            if self.q: break
        while self.q:
            n = len(self.q)
            for i in range(n):
                x, y = self.q.pop(0)
                for a, b in self.moves:
                    nx, ny = x + a, y + b
                    if nx < 0 or ny < 0 or nx >= self.r or ny >= self.c: continue
                    if self.visited[nx][ny]: continue
                    if self.grid[nx][ny] == 1: return count
                    self.visited[nx][ny] = True
                    self.q.append((nx, ny))
            count += 1
        return -1
        
    def find(self, x, y):
        if x < 0 or y < 0 or x >= self.r or y >= self.c: return
        if self.visited[x][y] or self.grid[x][y] == 0: return
        self.visited[x][y] = True
        self.q.append((x, y))
        for a, b in self.moves:
            self.find(x + a, y + b)
